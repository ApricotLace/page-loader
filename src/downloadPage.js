import axios from 'axios';
import { promises as fs } from 'fs';
import path from 'path';

const constructPath = (originalPath, url, isforFolder = 0) => {
  const parsedUrl = new URL(url);
  const rawNname = parsedUrl.href.slice(`${parsedUrl.protocol}//`.length);
  const cebabNotation = `${rawNname.match(/\w+/g).join('-')}`;
  const nameInCebabNotation = isforFolder ? `${cebabNotation}_files` : `${cebabNotation}.html`;
  return path.join(originalPath, nameInCebabNotation);
};

export default (downloadPath, downloadedResoursePath) => axios.get(downloadedResoursePath)
  .then((response) => {
    const readyPath = constructPath(downloadPath, response.config.url);
    return fs.writeFile(readyPath, response.data)
      .then(() => readyPath);
  })
  .catch((err) => {
    console.log(`${err}`);
    throw err;
  });
