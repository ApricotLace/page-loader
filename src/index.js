import axios from 'axios';
import { promises as fs } from 'fs';
import path from 'path';

const constructPath = (originalPath, url) => {
  const parsedUrl = new URL(url);
  const { protocol } = parsedUrl;
  const rawFilename = parsedUrl.href.slice(`${protocol}//`.length);
  const filenameInCebabNotation = `${rawFilename.match(/\w+/g).join('-')}.html`;
  return path.join(originalPath, filenameInCebabNotation);
};
export default
(downloadPath, downloadedResoursePath) => axios.get(downloadedResoursePath)
  .then((response) => {
    const readyPath = constructPath(downloadPath, response.config.url);
    return fs.writeFile(readyPath, response.data)
      .then(() => readyPath)
      .catch(err => console.log(`${err}`));
  })
  .catch(err => console.log(`${err}`));
