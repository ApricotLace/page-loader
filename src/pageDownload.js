import axios from 'axios';
import { promises as fs } from 'fs';
import path from 'path';

export const constructPath = (originalPath, url, isForResFolder = 0) => {
  const parsedUrl = new URL(url);
  const rawNname = parsedUrl.href.slice(`${parsedUrl.protocol}//`.length);
  const cebabNotation = `${rawNname.match(/\w+/g).join('-')}`;
  const nameInCebabNotation = isForResFolder
    ? `${cebabNotation}_files` : `${cebabNotation}.html`;
  return path.join(originalPath, nameInCebabNotation);
};

export const download = (downloadPath, downloadedResoursePath) => {
  const readyPath = constructPath(downloadPath, downloadedResoursePath);
  return axios.get(downloadedResoursePath)
    .then(response => fs.writeFile(readyPath, response.data))
    .then(() => readyPath);
};
