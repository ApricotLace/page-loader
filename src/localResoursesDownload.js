import cheerio from 'cheerio';
import fs from 'fs';
import path from 'path';
import url from 'url';
import axios from 'axios';
import { constructPath } from './pageDownload';

const mapping = {
  img: 'src',
  script: 'src',
  link: 'href',
};

const getAttr = tag => tag.attribs[mapping[tag.name]];

const predicate = (tag) => {
  const attr = getAttr(tag);
  if (!attr || tag.attribs.rel === 'canonical') {
    return false;
  }
  if (attr.substr(0, 2) === '//') {
    return false;
  }
  return !url.parse(attr).host;
};


const getResoursesDirName = sourceLink => constructPath('', sourceLink, 1);

const constructFileName = (pathToRes) => {
  const fileExt = path.extname(pathToRes);
  if (fileExt) {
    const fixedFileExt = fileExt.includes('?') ? fileExt.substr(0, fileExt.indexOf('?')) : fileExt;
    const rawFileName = pathToRes.slice(1, pathToRes.indexOf(fixedFileExt));
    return `${rawFileName.match(/\w+/g).join('-')}${fixedFileExt}`;
  }
  const rawFileName = pathToRes.slice(1);
  return `${rawFileName.match(/\w+/g).join('-')}`;
};

const updateHtml = (html, sourceLink, pathToPage) => {
  const resDirName = getResoursesDirName(sourceLink);
  const parsedUrl = new URL(sourceLink);

  const $ = cheerio.load(html);
  const arrayOfTagsObjects = [$('link'), $('img'), $('script')];
  const pathsToResourses = arrayOfTagsObjects
    .reduce((acc, tagsObj) => {
      const filteredTags = tagsObj
        .filter((i, tag) => predicate(tag));
      $(filteredTags).each((i, tag) => {
        acc.push(`${parsedUrl.origin}${getAttr(tag)}`);
        $(tag).attr(`${mapping[tag.name]}`, `${resDirName}/${constructFileName(getAttr(tag))}`);
      });
      return acc;
    }, []);
  return fs.promises.writeFile(pathToPage, $.html())
    .then(() => pathsToResourses);
};

const downloadLocalRes = (links, pathToResDir) => {
  const getPathName = (response) => {
    const parsedUrl = new URL(response.config.url);
    return parsedUrl.pathname;
  };
  const promises = links.map(link => axios.get(link, { responseType: 'stream' }));
  return Promise.all(promises)
    .then(([...data]) => data)
    .then(responses => responses
      .forEach(response => response
        .data.pipe(fs.createWriteStream(`${pathToResDir}/${constructFileName(getPathName(response))}`))));
};

export default (pathToPage, sourceLink) => {
  const pathToResDir = `${pathToPage.slice(0, pathToPage.indexOf('.html'))}_files`;
  return fs.promises.mkdir(pathToResDir)
    .then(() => fs.promises.readFile(pathToPage))
    .then(data => updateHtml(data.toString(), sourceLink, pathToPage))
    .then(links => downloadLocalRes(links, pathToResDir))
    .then(() => pathToResDir);
};
