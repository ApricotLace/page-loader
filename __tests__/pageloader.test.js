import axios from 'axios';
import httpAdapter from 'axios/lib/adapters/http';
import nock from 'nock';
import { promises as fs } from 'fs';
import path from 'path';
import os from 'os';
import { download, loadPage } from '../src';

axios.defaults.adapter = httpAdapter;
const getPathToFixture = fixtureName => `${__dirname}/__fixtures__/${fixtureName}`;

describe('download test', () => {
  it('#succ 200', async () => {
    const host = 'https://localhost';
    nock(host)
      .get('/')
      .reply(200, 'test data');

    const response = await axios.get(host);
    expect(response.data).toBe('test data');
  });
  it('#succ download', async () => {
    const host = 'https://localhost';
    nock(host)
      .get('/')
      .reply(200, 'some html code goes here\nalso a lot of tags');

    const tmpFilePath = await fs.mkdtemp(path.join(os.tmpdir(), 'fancytest'));
    const pathToWrittenFile = await download(tmpFilePath, host);
    const writtenFile = await fs.readFile(pathToWrittenFile);
    expect(writtenFile.toString()).toBe('some html code goes here\nalso a lot of tags');
  });
});

describe('write&update test', () => {
  const link = 'http://localhost';
  beforeEach(() => {
    nock(link)
      .get('/')
      .replyWithFile(200, getPathToFixture('originalHtmlTest2.html'));

    nock(link)
      .get('/cdn-cgi/scripts/5c5dd728/cloudflare-static/email-decode.min.js')
      .replyWithFile(200, getPathToFixture('testLocalRes'));

    nock(link)
      .get('/courses')
      .replyWithFile(200, getPathToFixture('originalHtmlTest2.html'));
  });
  it('#succ local download', async () => {
    const pathToTmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'hash'));
    const pathToResDir = await loadPage(pathToTmpDir, link);
    const dirContent = await fs.readdir(pathToResDir[1]);
    expect(dirContent.length).toBe(2);
  });

  it('#succ html update', async () => {
    const pathToTmpDir = await fs.mkdtemp(path.join(os.tmpdir(), 'hash'));
    await loadPage(pathToTmpDir, link);
    const fixture = await fs.readFile(getPathToFixture('originalHtmlTest2.html'));
    const dirContent = await fs.readdir(pathToTmpDir);
    const writtenFile = await fs.readFile(`${pathToTmpDir}/${dirContent[0]}`);
    expect(writtenFile.toString()).not.toBe(fixture.toString());
  });
});
