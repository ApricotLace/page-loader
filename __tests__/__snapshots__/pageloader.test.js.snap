// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`localResDownload fail #fails 1`] = `"Request failed with status code 404"`;

exports[`rawPageDownload fail #fails 1`] = `"Request failed with status code 404"`;
